src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/InMemoryPartitionedUserTokenCacheAccessor.cs

```diff
@@ -238,5 +244,16 @@ public virtual void Clear()
             AccountCacheDictionary.Clear();
             // app metadata isn't removable
         }
+
+        public bool HasAccessOrRefreshTokens()
+        {
+            return RefreshTokenCacheDictionary.Any(partition => partition.Value.Count > 0) ||
+                    AccessTokenCacheDictionary.Any(partition => partition.Value.Any(token => !IsAtExpired(token.Value)));
+        }
+
+        private bool IsAtExpired(MsalAccessTokenCacheItem at)
+        {
+            return at.ExpiresOn < DateTime.UtcNow + Internal.Constants.AccessTokenExpirationBuffer;
```

Comment:
Logic should be on MsalAccessTokenCacheItem

Reply:
Moved.